content/Improve your Rails App searches with Elasticsearch - Leandro Dominguez.md

```diff
@@ -0,0 +1,216 @@
+
+Given a simplified `Post` model with only a `title`, `body` and a `topic`, include the next lines into it:
+
+```ruby
+class Post < ApplicationRecord
+ include Elasticsearch::Model
+ include Elasticsearch::Model::Callbacks
+end
+```
+
+The first `include` will let us make Elasticsearch calls in a friendly ActiveRecord-like syntax. This way we can, for example, just call `Post.search(query)` for searching. Otherwise we should manually make an API call to the Elastic server instance every time we want to interact with our data.
+
+The second `include` adds callbacks in our model, so every time we create, delete or edit a `Post`, this action will be replicated into Elasticsearch data.
+
+This last point is very important. Take into account we have two independent databases: one in which we’re gonna persist all the model data as you’re used to (Postgres) and this other new one for making fast searches, suggestions and use all other Elasticsearch fantastic features.
+
+Sometimes (for performance or even business requirements), we don’t want to update data in Elasticsearch instantaneously every time a record has been changed in our main database (this is customizable but we won’t tackle this in the current article).
+
+**Step 3: Defining our Elasticsearch index**
+
+In Elasticsearch we don’t have tables or rows. Our models are stored in _`indexes`_ and each instance is called _`document`_. Although Elastic can automatically detect the attributes types of our model, we are going to declare them explicitly so we have more control over them by adding the following to your model:
+
+```ruby
+ settings do
+   mappings dynamic: false do
+     indexes :id, index: :not_analyzed
+     indexes :title, type: :text, analyzer: :english
+     indexes :body, type: :text, analyzer: :english
+     indexes :topic, type: :keyword
+   end
+ end
+```
+
+**What are we doing here?**
+
+In the second line, with `dynamic: false` we are avoiding new fields to be created accidentally. If you choose `dynamic: true`, every not-listed attribute will be automatically added in the index in an update or create query.
+
+In the third line we are telling Elastic we want to take our `id` attribute and add it to the `Post`’ index as an index but without analyzing it. This means it won’t apply any filter or transformation to it.
+
+In the fourth and fifth line, we're adding both `title` and `body` as text type and with an English analyzer. This will automatically run text tokenizers, analyzers and will apply a lot of natural language processing tools every time we insert or edit a document in our index, making the searches faster, easier and typo-resistant (be aware all these things take an extra time). Follow these links if you’re interested in learning more about [language analyzers](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-lang-analyzer.html) and [tokenizers](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-tokenizers.html).
```

Comment:
```suggestion
In the fourth and fifth lines, we're adding both `title` and `body` as text type and with an English analyzer. This will automatically run text tokenizers, analyzers and will apply a lot of natural language processing tools every time we insert or edit a document in our index, making the searches faster, easier, and typo-resistant (be aware all these things take extra time). Follow these links if you’re interested in learning more about [language analyzers](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-lang-analyzer.html) and [tokenizers](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-tokenizers.html).
```